### PR TITLE
Enable dynamic membership

### PR DESCRIPTION
**Pull request by derekchiang**
This PR fixes a couple bugs that have been preventing dynamic membership (i.e. the ability to add/remove nodes without affecting the availability of the cluster) from working.

### CODE REVIEW

File: etc/kube/pachyderm.json [hunk 1]
```diff
@@ -343,7 +343,10 @@
       "spec": {
         "volumes": [
           {
-            "name": "pach-disk"
+            "name": "pach-disk",
+            "hostPath": {
+              "path": "/tmp/pach-402374041"
```
Comment (jdoliner) [inline etc/kube/pachyderm.json]: What does this number signify?

Comment (derekchiang) [inline etc/kube/pachyderm.json]: It's generated from pach-deploy.  It's just a random directory.  Running pach-deploy again will result in a different directory.

File: src/server/pfs/server/api_server.go [hunk 2]
```diff
@@ -704,9 +717,27 @@ func (a *apiServer) DeleteFile(ctx context.Context, request *pfs.DeleteFileReque
...
+		defer a.versionLock.RUnlock()
+
+		a.versionChanLock.Lock()
+		defer a.versionChanLock.Unlock()
+
+		if !a.versionChanClosed[a.version] {
+			close(a.versionChans[a.version])
+			a.versionChanClosed[a.version] = true
+			a.versionChans[version] = make(chan struct{})
+		}
```
Comment (jdoliner) [inline src/server/pfs/server/api_server.go]: I don't think you need the `versionChanClosed` map because I think `Version` should only get called once with each value.

File: src/server/pkg/ctx/context.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	if v2 != nil {
+		return v2
+	}
+
+	return nil
+}
+
+func (c *Context) Finish() {
+	close(c.finish)
+}
```
Comment (jdoliner) [inline src/server/pkg/ctx/context.go]: Is this file still necessary.

File: src/server/pfs/server/api_server.go [hunk 4]
```diff
@@ -725,9 +756,27 @@ func (a *apiServer) getClientConnForFile(file *pfs.File, version int64) (*grpc.C
...
+		case <-done:
+		}
+	}()
+	ctx = metadata.NewContext(
 		ctx,
-		metadata.Pairs("version", fmt.Sprint(version)),
+		metadata.Pairs("version", fmt.Sprint(a.version)),
 	)
+	return ctx, done
 }
```
Comment (jdoliner) [inline src/server/pfs/server/api_server.go]: Looks good right to me.

**Comment by derekchiang** [comment]
To be clear, this PR merely ensures that the cluster will be **functional** after membership changes.  Outstanding requests are still going to be interrupted when membership changes, which is very much less than ideal, especially considering the fact that we currently don't have a way to restart a pipeline after a fails.

**Comment by derekchiang** [comment]
@jdoliner could you review?

File: src/server/pachyderm_test.go [hunk 5]
```diff
@@ -1544,7 +1544,7 @@ func TestClusterFunctioningAfterMembershipChange(t *testing.T) {
 	scalePachd(t, k)
 	// Wait for the cluster to stablize... ideally we shouldn't have to
 	// do that.
-	time.Sleep(30 * time.Second)
+	time.Sleep(20 * time.Second)
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: Is there a way to wait for this with the kube client?

**Comment by jdoliner** [comment]
This LGTM after the temp dir comment and hopefully the comment about the test sleep are addressed.
The later might be too much work but the former I'd like to see.

File: src/server/pkg/deploy/assets/assets.go [hunk 6]
```diff
@@ -80,6 +81,16 @@ func PachdRc(shards uint64, backend backend) *api.ReplicationController {
 	var backendEnvVar string
 	switch backend {
 	case localBackend:
+		if hostPath == "" {
+			dir, err := ioutil.TempDir("/tmp", "pach-")
```
Comment (derekchiang) [inline src/server/pkg/deploy/assets/assets.go]: But what if `/tmp/pach/` doesn't exist?  Does our code address that?

Comment (jdoliner) [inline src/server/pkg/deploy/assets/assets.go]: Does this code address that? All this will do is guarantee that the machine that generates the manifest has that directory. The machine that generates the manifest is seldom the machine that runs pachd since most people use the statically generated manifests in the repo. I think for us to use a hostpath in k8s we have to be assuming that k8s will create the directory if it doesn't exist.

**Comment by jdoliner** [comment]
1bf5jzd

**Comment by jdoliner** [comment]
Ignore the second part of that gif.

**Comment by derekchiang** [comment]
lmao
